### ISSUE TITLE
[CI] ProfileCancellationIntegTests testSuggestProfilesCancellation failing

### ISSUE THREAD

**Issue by ywelsch** [elastic/elasticsearch#86881]
**Build scan:**
https://gradle-enterprise.elastic.co/s/k3titqku4p2be/tests/:x-pack:plugin:security:internalClusterTest/org.elasticsearch.xpack.security.profile.ProfileCancellationIntegTests/testSuggestProfilesCancellation

**Reproduction line:**
`./gradlew ':x-pack:plugin:security:internalClusterTest' --tests "org.elasticsearch.xpack.security.profile.ProfileCancellationIntegTests.testSuggestProfilesCancellation" -Dtests.seed=74F6D1B80FACB410 -Dtests.locale=no-NO -Dtests.timezone=BET -Druntime.java=17`

**Applicable branches:**
master

**Reproduces locally?:**
No

**Failure history:**
https://gradle-enterprise.elastic.co/scans/tests?tests.container=org.elasticsearch.xpack.security.profile.ProfileCancellationIntegTests&tests.test=testSuggestProfilesCancellation

**Failure excerpt:**
```
java.lang.AssertionError: 
Expected: an empty collection
     but: <[107]>

  at __randomizedtesting.SeedInfo.seed([74F6D1B80FACB410:B026926B1FBFB102]:0)
  at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
  at org.junit.Assert.assertThat(Assert.java:956)
  at org.junit.Assert.assertThat(Assert.java:923)
  at org.elasticsearch.xpack.security.profile.ProfileCancellationIntegTests.lambda$testSuggestProfilesCancellation$3(ProfileCancellationIntegTests.java:119)
  at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:1096)
  at org.elasticsearch.xpack.security.profile.ProfileCancellationIntegTests.testSuggestProfilesCancellation(ProfileCancellationIntegTests.java:109)
  at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(NativeMethodAccessorImpl.java:-2)
  at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
  at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
  at java.lang.reflect.Method.invoke(Method.java:568)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1758)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:946)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:982)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:996)
  at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:44)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:45)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:375)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:824)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:475)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:955)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:840)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:891)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:902)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:38)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:47)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:375)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.lambda$forkTimeoutingTask$0(ThreadLeakControl.java:831)
  at java.lang.Thread.run(Thread.java:833)

```

**Comment by n1v0lg** [comment]
One odd thing I noticed in the logs is that `blocking search on` never gets logged. I wonder if inside `enableSearchBlock`  not all instances of `PluginsService` are available yet when the method gets called and so the block plugin doesn't even get registered.

**Comment by ywangd** [comment]
Thanks for looking into this issue.

> One odd thing I noticed in the logs is that blocking search on never gets logged for the failures

I also noticed the same thing last time this test failed. My thinking was that the cluster was slow at handling requests so that the shardSearchRequest didn't get handled in time (the message will only be logged then). So I expanded the assertion to match 3 tasks: suggest, search and shardSearch to make sure they are all handled before calling cancel. But it seems not enough. Now we are sure all tasks are registered as expected. But that does not mean they are actually handled because the message is not logged. Whenever the original task did not get actually handled, cancellation does not work expected. I suspect underlying they are still cancelled. It's just the state is not exactly as expected. Or maybe the task just gets lingered for a long while. In this particular failure, the search request was not cancelled. But both of its parent (suggest) and child (shardSearch) did get cancelled. Maybe it was waiting for something. I need dig a bit more.

> I wonder if inside enableSearchBlock not all instances of PluginsService are available yet when the method gets called and so the block plugin doesn't even get registered.

I don't think it's possible. First, the plugin is registerd (as show by the log line "loaded plugin ... BlockPlugin"). Second, the test suite makes sure the test is in an usable state before each test method so all components are in place before `enableSearchBlock`.
